src/Schema/Directives/AllDirective.php

```diff
@@ -56,13 +58,48 @@ public function resolveField(FieldValue $fieldValue): FieldValue
                 $query = $this->getModelClass()::query();
             }
 
-            return $resolveInfo
+            $builder = $resolveInfo
                 ->argumentSet
                 ->enhanceBuilder(
                     $query,
                     $this->directiveArgValue('scopes', [])
-                )
-                ->get();
+                );
+
+            if (config('lighthouse.optimized_selects')) {
+                if (($builder instanceof QueryBuilder || $builder instanceof EloquentBuilder) && ! $this->directiveHasArgument('builder')) {
+                    $fieldSelection = array_keys($resolveInfo->getFieldSelection(1));
+
+                    $selectColumns = SelectHelper::getSelectColumns(
+                        $this->definitionNode,
+                        $fieldSelection,
+                        $this->getModelClass()
+                    );
+
+                    if (empty($selectColumns)) {
+                        return $builder->get();
+                    }
+
+                    $query = $builder instanceof EloquentBuilder ? $builder->getQuery() : $builder;
+
+                    if (null !== $query->columns) {
+                        $bindings = $query->getRawBindings();
+
+                        $expressions = array_filter($query->columns, function ($column) {
+                            return $column instanceof Expression;
+                        });
+
+                        $builder = $builder->select(array_unique(array_merge($selectColumns, $expressions)));
+
+                        foreach ($bindings as $type => $binding) {
+                            $builder = $builder->addBinding($binding, $type);
+                        }
```

Comment:
What does this do?

Reply:
We only want to replace the `*` with `id`. The sub-query and bindings should be preserved.

From `testOrderByRelationCount()`
```graphql
{
  users(
     orderBy: [
       {
          tasks: { aggregate: COUNT }
          order: ASC
       }
     ]
  ) {
    id
  }
}
```

query dump:
```php
columns: array:2 [
    0 => "users.*"
    1 => Illuminate\Database\Query\Expression^ {#7717
      #value: "(select count(*) from `tasks` where `users`.`id` = `tasks`.`user_id` and `tasks`.`deleted_at` is null and `name` != ?) as `tasks_count`"
    }
  ],
bindings: array:9 [
    "select" => array:1 [
      0 => "cleaning"
    ]
    "from" => []
    "join" => []
    "where" => []
    "groupBy" => []
    "having" => []
    "order" => []
    "union" => []
    "unionOrder" => []
  ]
```

Reply:
I do not understand why you need to first get and then re-set the bindings. As far as I can tell, `->select()` only clears `columns` and `bindings['select']`.

Reply:
Do you mean it only needs to re-set the `bindings['select']` but not the whole `bindings`?

Reply:
I think these 3 lines and `$bindings = $query->getRawBindings();` are completely unnecessary.

Reply:
The above schema is the case. Before `->select()`, the raw query is below. The subquery has a select binding `cleaning` that `name != ?` needs. If we don't re-set the bindings, it will throw `Illuminate\Database\QueryException: SQLSTATE[HY093]: Invalid parameter number`. Because the select bindings were clear by `->select()`
```sql
SELECT 
    `users.*`,
    (SELECT 
            COUNT(*)
        FROM
            `tasks`
        WHERE
            `users`.`id` = `tasks`.`user_id`
                AND `tasks`.`deleted_at` IS NULL
                AND `name` != ?
	) AS `tasks_count`
FROM
    `users`
ORDER BY `tasks_count` ASC
```

Reply:
> Do you mean it only needs to re-set the bindings['select'] but not the whole bindings?

That sounds reasonable then. Saves unnecessary work and is more explicit about what is happening and why.